Flaky test in 1.2 branch: TestStatsPagination_RSC6a_RSCb3

Sometimes all 3 test scenarios fail, right now I am getting this error 
```
=== RUN   TestStatsPagination_RSC6a_RSCb3
=== PAUSE TestStatsPagination_RSC6a_RSCb3
=== CONT  TestStatsPagination_RSC6a_RSCb3
--- FAIL: TestStatsPagination_RSC6a_RSCb3 (0.00s)
=== RUN   TestStatsPagination_RSC6a_RSCb3/limit=2
=== PAUSE TestStatsPagination_RSC6a_RSCb3/limit=2
=== CONT  TestStatsPagination_RSC6a_RSCb3/limit=2
    --- FAIL: TestStatsPagination_RSC6a_RSCb3/limit=2 (12.28s)

=== RUN   TestStatsPagination_RSC6a_RSCb3/limit=3
=== PAUSE TestStatsPagination_RSC6a_RSCb3/limit=3
=== CONT  TestStatsPagination_RSC6a_RSCb3/limit=3
    TestStatsPagination_RSC6a_RSCb3/limit=20: rest_client_test.go:620: expected items: [<Stats 2020-01-28:14:09; unit=minute; count=0> <Stats 2020-01-28:14:08; unit=minute; count=0> <Stats 2020-01-28:14:07; unit=minute; count=0> <Stats 2020-01-28:14:06; unit=minute; count=0> <Stats 2020-01-28:14:05; unit=minute; count=0> <Stats 2020-01-28:14:04; unit=minute; count=0> <Stats 2020-01-28:14:03; unit=minute; count=0> <Stats 2020-01-28:14:02; unit=minute; count=0> <Stats 2020-01-28:14:01; unit=minute; count=0> <Stats 2020-01-28:14:00; unit=minute; count=0>], got: [<Stats 2021-05-14:12:08; unit=minute; count=0> <Stats 2020-01-28:14:09; unit=minute; count=0> <Stats 2020-01-28:14:08; unit=minute; count=0> <Stats 2020-01-28:14:07; unit=minute; count=0> <Stats 2020-01-28:14:06; unit=minute; count=0> <Stats 2020-01-28:14:05; unit=minute; count=0> <Stats 2020-01-28:14:04; unit=minute; count=0> <Stats 2020-01-28:14:03; unit=minute; count=0> <Stats 2020-01-28:14:02; unit=minute; count=0> <Stats 2020-01-28:14:01; unit=minute; count=0> <Stats 2020-01-28:14:00; unit=minute; count=0>]
    TestStatsPagination_RSC6a_RSCb3/limit=2: rest_client_test.go:620: expected items: [<Stats 2020-01-28:14:09; unit=minute; count=0> <Stats 2020-01-28:14:08; unit=minute; count=0> <Stats 2020-01-28:14:07; unit=minute; count=0> <Stats 2020-01-28:14:06; unit=minute; count=0> <Stats 2020-01-28:14:05; unit=minute; count=0> <Stats 2020-01-28:14:04; unit=minute; count=0> <Stats 2020-01-28:14:03; unit=minute; count=0> <Stats 2020-01-28:14:02; unit=minute; count=0> <Stats 2020-01-28:14:01; unit=minute; count=0> <Stats 2020-01-28:14:00; unit=minute; count=0>], got: [<Stats 2021-05-14:12:08; unit=minute; count=0> <Stats 2020-01-28:14:09; unit=minute; count=0> <Stats 2020-01-28:14:08; unit=minute; count=0> <Stats 2020-01-28:14:07; unit=minute; count=0> <Stats 2020-01-28:14:06; unit=minute; count=0> <Stats 2020-01-28:14:05; unit=minute; count=0> <Stats 2020-01-28:14:04; unit=minute; count=0> <Stats 2020-01-28:14:03; unit=minute; count=0> <Stats 2020-01-28:14:02; unit=minute; count=0> <Stats 2020-01-28:14:01; unit=minute; count=0> <Stats 2020-01-28:14:00; unit=minute; count=0>]
    --- PASS: TestStatsPagination_RSC6a_RSCb3/limit=3 (11.92s)
=== RUN   TestStatsPagination_RSC6a_RSCb3/limit=20
=== PAUSE TestStatsPagination_RSC6a_RSCb3/limit=20
=== CONT  TestStatsPagination_RSC6a_RSCb3/limit=20
    --- FAIL: TestStatsPagination_RSC6a_RSCb3/limit=20 (12.00s)


FAIL
```

--- a/ably/ablytest/pagination.go
+++ b/ably/ablytest/pagination.go
@@ -10,7 +10,7 @@ import (
 // request into the slice pointed to by dst, which must be a pointer to a slice
 // of the same type as the paginated response's items.
 func AllPages(dst, paginatedRequest interface{}) error {
-	_, getItems := generalizePagination(reflect.ValueOf(paginatedRequest))
+	_, getItems := generalizePaginatedRequest(reflect.ValueOf(paginatedRequest))
 	items, err := getItems()
 	if err != nil {
 		return err
@@ -41,6 +41,10 @@ func PaginationWithSortResult(sort func([]interface{})) PaginationOption {
 	}
 }
 
+// TestPagination takes a PaginatedRequest and the unwrapped expected items its
+// resulting pages should contain, and tests that the pages do contain those
+// items, plus some generic PaginatedResponse functionality, such as Pages and
+// Items methods, and the First method.
 func TestPagination(expected, request interface{}, perPage int, options ...PaginationOption) error {
 	opts := paginationOptions{
 		equal:      reflect.DeepEqual,
@@ -59,7 +63,7 @@ func TestPagination(expected, request interface{}, perPage int, options ...Pagin
 }
 
 func testPagination(request reflect.Value, expectedItems []interface{}, perPage int, opts paginationOptions) error {
-	getPages, getItems := generalizePagination(request)
+	getPages, getItems := generalizePaginatedRequest(request)
 
 	for i := 0; i < 2; i++ {
 		pages, err := getPages()
@@ -135,7 +139,17 @@ type paginatedItems struct {
 	item func() interface{}
 }
 
-func generalizePagination(request reflect.Value) (func() (paginatedResult, error), func() (paginatedItems, error)) {
+// generalizePaginatedRequest takes a non-generic PaginationRequest and returns
+// functions that call its Pages and Items methods, mapping paginated items
+// to interface{}.
+//
+// This can be used both to test that the provided PaginationRequest implements
+// the informal PaginationRequest interface correctly and that the resulting
+// paginated items are the ones we expect.
+func generalizePaginatedRequest(request reflect.Value) (
+	pages func() (paginatedResult, error),
+	items func() (paginatedItems, error),
+) {
 	ctx := reflect.ValueOf(context.Background())
 
 	generalizeCommon := func(r reflect.Value) paginated {
@@ -154,7 +168,7 @@ func generalizePagination(request reflect.Value) (func() (paginatedResult, error
 		}
 	}
 
-	pages := func() (paginatedResult, error) {
+	pages = func() (paginatedResult, error) {
 		ret := request.MethodByName("Pages").Call([]reflect.Value{ctx})
 		if err, ok := ret[1].Interface().(error); ok && err != nil {
 			return paginatedResult{}, err
@@ -173,7 +187,7 @@ func generalizePagination(request reflect.Value) (func() (paginatedResult, error
 		}, nil
 	}
 
-	items := func() (paginatedItems, error) {
+	items = func() (paginatedItems, error) {
 		ret := request.MethodByName("Items").Call([]reflect.Value{ctx})
 		if err, ok := ret[1].Interface().(error); ok && err != nil {
 			return paginatedItems{}, err

--- a/ably/rest_client_test.go
+++ b/ably/rest_client_test.go
@@ -613,7 +613,13 @@ func TestStatsPagination_RSC6a_RSCb3(t *testing.T) {
 
 			err := ablytest.TestPagination(
 				reverseStats(fixtures),
-				rest.Stats(ably.StatsWithLimit(limit)),
+				rest.Stats(
+					ably.StatsWithLimit(limit),
+
+					// We must set an end parameter. Otherwise, we may get the
+					// *current* minute's stats alongside the fixtures.
+					ably.StatsWithEnd(time.Date(2020, time.January, 29, 0, 0, 0, 0, time.UTC)),
+				),
 				limit,
 			)
 			if err != nil {
@@ -688,6 +694,10 @@ func TestStats_Direction_RSC6b2(t *testing.T) {
 			pages, err := rest.Stats(
 				ably.StatsWithLimit(len(expected)),
 				ably.StatsWithDirection(c.direction),
+
+				// We must set an end parameter. Otherwise, we may get the
+				// *current* minute's stats alongside the fixtures.
+				ably.StatsWithEnd(time.Date(2020, time.January, 29, 0, 0, 0, 0, time.UTC)),
 			).Pages(ctx)
 			if err != nil {
 				t.Fatal(err)
@@ -720,6 +730,10 @@ func TestStats_Unit_RSC6b4(t *testing.T) {
 
 	pages, err := rest.Stats(
 		ably.StatsWithUnit(ably.PeriodMonth),
+
+		// We must set an end parameter. Otherwise, we may get the
+		// *current* minute's stats alongside the fixtures.
+		ably.StatsWithEnd(time.Date(2020, time.January, 29, 0, 0, 0, 0, time.UTC)),
 	).Pages(ctx)
 	if err != nil {
 		t.Fatal(err)

